tests/contrib/asgi/test_asgi.py

```diff
@@ -0,0 +1,261 @@
+import asyncio
+from functools import partial
+import random
+import sys
+
+import httpx
+import pytest
+from asgiref.testing import ApplicationCommunicator
+from ddtrace.contrib.asgi import TraceMiddleware
+from ddtrace.propagation import http as http_propagation
+from tests.base import BaseTestCase
+from tests.test_tracer import get_dummy_tracer
+
+
+@pytest.fixture(
+    params=[
+        {},
+        {"server": None},
+        {"server": ("dev", 8000)},
+        {"http_version": "1.0", "asgi": {}},
+        {"http_version": "1.0", "asgi": {"version": "3.2"}},
+        {"http_version": "1.0", "asgi": {"spec_version": "2.1",}},
+        {"http_version": "1.0", "asgi": {"version": "3.2", "spec_version": "2.1",}},
+    ]
+)
+def scope(request):
+    s = {
+        "client": ("127.0.0.1", 32767),
+        "headers": [],
+        "method": "GET",
+        "path": "/",
+        "query_string": b"",
+        "scheme": "http",
+        "server": ("127.0.0.1", 80),
+        "type": "http",
+    }
+    s.update(request.param)
+    return {k: v for (k, v) in s.items() if v is not None}
+
+
+@pytest.fixture
+def tracer():
+    tracer = get_dummy_tracer()
+    if sys.version_info < (3, 7):
+        # enable legacy asyncio support
+        from ddtrace.contrib.asyncio.provider import AsyncioContextProvider
+
+        tracer.configure(context_provider=AsyncioContextProvider())
+    yield tracer
+
+
+async def basic_app(scope, receive, send):
+    message = await receive()
+    assert scope["type"] == "http"
```

Comment:
(Nit) Generally best practice to verify the scope type _before_ interacting with the server - so just to get this point across…

```suggestion
    assert scope["type"] == "http"
    message = await receive()
```